File: core/core-web/src/main/java/ch/difty/scipamato/core/web/user/UserEditPage.java
package ch.difty.scipamato.core.web.user;

import static ch.difty.scipamato.core.entity.User.UserFields.*;

import java.util.Arrays;
import java.util.Optional;

import de.agilecoders.wicket.core.markup.html.bootstrap.button.BootstrapButton;
import de.agilecoders.wicket.core.markup.html.bootstrap.button.Buttons;
import de.agilecoders.wicket.extensions.markup.html.bootstrap.confirmation.ConfirmationBehavior;
import de.agilecoders.wicket.extensions.markup.html.bootstrap.form.select.BootstrapMultiSelect;
import de.agilecoders.wicket.extensions.markup.html.bootstrap.form.select.BootstrapSelectConfig;
import lombok.extern.slf4j.Slf4j;
import org.apache.wicket.AttributeModifier;
import org.apache.wicket.authorization.UnauthorizedInstantiationException;
import org.apache.wicket.authroles.authorization.strategies.role.annotations.AuthorizeInstantiation;
import org.apache.wicket.bean.validation.PropertyValidator;
import org.apache.wicket.markup.html.basic.Label;
import org.apache.wicket.markup.html.form.*;
import org.apache.wicket.markup.html.form.validation.EqualPasswordInputValidator;
import org.apache.wicket.model.CompoundPropertyModel;
import org.apache.wicket.model.Model;
import org.apache.wicket.model.PropertyModel;
import org.apache.wicket.model.StringResourceModel;
import org.apache.wicket.request.mapper.parameter.PageParameters;
import org.apache.wicket.spring.injection.annot.SpringBean;
import org.apache.wicket.util.string.StringValue;
import org.jetbrains.annotations.NotNull;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.wicketstuff.annotation.mount.MountPath;

import ch.difty.scipamato.core.auth.Role;
import ch.difty.scipamato.core.auth.Roles;
import ch.difty.scipamato.core.entity.User;
import ch.difty.scipamato.core.persistence.OptimisticLockingException;
import ch.difty.scipamato.core.persistence.UserService;
import ch.difty.scipamato.core.web.CorePageParameters;
import ch.difty.scipamato.core.web.common.BasePage;

/**
 * Loads the setting of the user with the id supplied in the page parameters.
 * If no valid integer is supplied in the page parameters, the settings of the
 * user in the current session is presented.
 * <p>
 * The page can be opened in different modes:
 * <ul>
 * <li><b>CREATE:</b> The admin can create a new user</li>
 * <li><b>MANAGE:</b> The admin edits other users accounts (including passwords)</li>
 * <li><b>EDIT:</b> The active user can edit parts of his/her profile but cannot change the password.</li>
 * <li><b>CHANGE_PASSWORD:</b> The (active) user can change the password but no other aspects of his/her profile</li>
 * </ul>
 */
@MountPath("user")
@Slf4j
@AuthorizeInstantiation({ Roles.USER, Roles.ADMIN })
@SuppressWarnings({ "SameParameterValue", "WicketForgeJavaIdInspection" })
public class UserEditPage extends BasePage<ChangePasswordUser> {

    private static final long serialVersionUID = 1L;

    private static final String AM_DATA_WIDTH = "data-width";
    private static final String ROLES         = "roles";
    private static final String ROLES_STRING  = "rolesString";

    private final Mode mode;

    @SuppressWarnings("unused")
    @SpringBean
    private PasswordEncoder passwordEncoder;

    @SuppressWarnings("unused")
    @SpringBean
    private UserService userService;

    public UserEditPage(@NotNull final PageParameters pp) {
        super(pp);
        this.mode = getModeFrom(pp);
        assertManageOrCreateModeForAdminOnly();
        initDefaultModel(pp);
    }

    private Mode getModeFrom(final PageParameters pp) {
        final StringValue modeValue = pp.get(CorePageParameters.MODE.getName());
        return modeValue.isEmpty() ? Mode.EDIT : Mode.valueOf(modeValue.toString());
    }

    private void assertManageOrCreateModeForAdminOnly() {
        final User activeUser = getActiveUser();
        if (isInAdminMode() && !activeUser
            .getRoles()
            .contains(Role.ADMIN)) {
            final String msg = new StringResourceModel("page.access-forbidden", this, null)
                .setParameters(activeUser.getUserName(), this.getClass())
                .getString();
            log.warn(msg);
            throw new UnauthorizedInstantiationException(getPageClass());
        }
    }

    private boolean isInAdminMode() {
        return mode == Mode.MANAGE || mode == Mode.CREATE;
    }

    private boolean isInUserOwnedMode() {
        return !isInAdminMode();
    }

    private boolean isInPasswordResetMode() {
        return mode == Mode.CHANGE_PASSWORD;
    }

    private boolean canSetPasswords() {
        return isInAdminMode() || isInPasswordResetMode();
    }

    private void initDefaultModel(final PageParameters pp) {
        final StringValue userId = pp.get(CorePageParameters.USER_ID.getName());
        determineUserId(userId).ifPresentOrElse(this::loadUserWithIdFromDb,
            () -> setDefaultModel(Model.of(new ChangePasswordUser())));
    }

    /**
     * If the user id is supplied as page parameter, use its id (or that of the active user
     * if the parameter is not a numeric value). Otherwise use the id of the current user,
     * unless if in MANAGE Mode, then prepare for new user by returning an empty optional.
     */
    private Optional<Integer> determineUserId(final StringValue userIdStringValue) {
        final Integer activeUserId = getActiveUser().getId();
        switch (mode) {
        case CREATE:
            return Optional.empty();
        case MANAGE:
            return Optional.of(userIdStringValue.toInt(activeUserId));
        case EDIT:
        case CHANGE_PASSWORD:
        default:
            return Optional.of(activeUserId);
        }
    }

    private void loadUserWithIdFromDb(final int userId) {
        final boolean clearPassword = !isInPasswordResetMode();
        userService
            .findById(userId)
            .ifPresent(u -> setDefaultModel(Model.of(new ChangePasswordUser(u, clearPassword))));
    }

    @Override
    protected void onInitialize() {
        super.onInitialize();

        final Form<ChangePasswordUser> form = new Form<>("form", new CompoundPropertyModel<>(getModel()));
        queue(form);

        final TextField<String> userNameField = new TextField<>(USER_NAME.getFieldName());
        userNameField.setEnabled(isInAdminMode());
        queueFieldAndLabel(userNameField, new PropertyValidator<>());

        final TextField<String> firstNameField = new TextField<>(FIRST_NAME.getFieldName());
        firstNameField.setEnabled(!isInPasswordResetMode());
        queueFieldAndLabel(firstNameField, new PropertyValidator<>());

        final TextField<String> lastNameField = new TextField<>(LAST_NAME.getFieldName());
        lastNameField.setEnabled(!isInPasswordResetMode());
        queueFieldAndLabel(lastNameField, new PropertyValidator<>());

        final EmailTextField emailField = new EmailTextField(EMAIL.getFieldName());
        emailField.setEnabled(!isInPasswordResetMode());
        queueFieldAndLabel(emailField);

        final CheckBox enabledField = new CheckBox(ENABLED.getFieldName());
        enabledField.setEnabled(isInAdminMode());
        queueFieldAndLabel(enabledField);

        final Label rolesLabel = new Label(ROLES + LABEL_TAG,
            new StringResourceModel(ROLES + LABEL_RESOURCE_TAG, this, null));
        rolesLabel.setVisible(isInAdminMode());
        queue(rolesLabel);

        final BootstrapMultiSelect<Role> rolesMultiSelect = new BootstrapMultiSelect<>(ROLES,
            new PropertyModel<>(getModel(), ROLES), Arrays.asList(Role.values()), new EnumChoiceRenderer<>(this));
        rolesMultiSelect.with(new BootstrapSelectConfig()
            .withMultiple(true)
            .withLiveSearch(true)
            .withLiveSearchStyle("startsWith"));
        rolesMultiSelect.add(new AttributeModifier(AM_DATA_WIDTH, "auto"));
        rolesMultiSelect.setVisible(isInAdminMode());
        queue(rolesMultiSelect);

        final Label rolesStringLabel = new Label(ROLES_STRING + LABEL_TAG,
            new StringResourceModel(ROLES_STRING + LABEL_RESOURCE_TAG, this, null));
        rolesStringLabel.setVisible(isInUserOwnedMode());
        queue(rolesStringLabel);
        final Label rolesString = new Label(ROLES_STRING, getModelObject().getRolesString());
        rolesString.setVisible(isInUserOwnedMode());
        queue(rolesString);

        final PasswordTextField pwcField = new PasswordTextField("currentPassword");
        pwcField.setRequired(isInPasswordResetMode());
        pwcField.setVisible(isInPasswordResetMode());
        final String currentPassword = getModelObject().getPassword();
        if (isInPasswordResetMode() && currentPassword != null)
            pwcField.add(new CurrentPasswordMatchesValidator(passwordEncoder, currentPassword));
        queueFieldAndLabel(pwcField);

        final PasswordTextField pw1Field = new PasswordTextField(PASSWORD.getFieldName());
        pw1Field.setRequired(isInPasswordResetMode());
        pw1Field.setVisible(canSetPasswords());
        queueFieldAndLabel(pw1Field);

        final PasswordTextField pw2Field = new PasswordTextField("password2");
        pw2Field.setRequired(isInPasswordResetMode());
        pw2Field.setVisible(canSetPasswords());
        queueFieldAndLabel(pw2Field);

        form.add(new EqualPasswordInputValidator(pw1Field, pw2Field));

        queueSubmitButton("submit");

        final BootstrapButton deleteButton = new BootstrapButton("delete", new StringResourceModel("delete.label"),
            Buttons.Type.Default) {
            @Override
            public void onSubmit() {
                super.onSubmit();
                doOnDelete();
            }
        };
        deleteButton.add(new ConfirmationBehavior());
        queue(deleteButton);

    }

    private void queueSubmitButton(final String id) {
        queue(new BootstrapButton(id, new StringResourceModel("submit.label"), Buttons.Type.Default) {
            @Override
            public void onSubmit() {
                super.onSubmit();
                doOnSubmit();
            }
        });
    }

    private void doOnSubmit() {
        try {
            final User unsaved = getModelObject().toUser();
            if (!canSetPasswords())
                unsaved.setPassword(null);
            final User user = userService.saveOrUpdate(unsaved);
            if (user != null) {
                UserEditPage.this.setModelObject(new ChangePasswordUser(user));
                if (mode == Mode.CHANGE_PASSWORD) {
                    info(new StringResourceModel("password.changed.hint", this,
                        UserEditPage.this.getModel()).getString());

                } else {
                    info(new StringResourceModel("save.successful.hint", this, null)
                        .setParameters(getModelObject().getId(), user.getUserName())
                        .getString());
                }
            } else {
                error(new StringResourceModel("save.error.hint", this, null)
                    .setParameters(getModelObject().getId(), "")
                    .getString());
            }
        } catch (OptimisticLockingException ole) {
            final String msg = new StringResourceModel("save.optimisticlockexception.hint", this, null)
                .setParameters(ole.getTableName(), getModelObject().getId())
                .getString();
            log.error(msg);
            error(msg);
        } catch (Exception ex) {
            String msg = new StringResourceModel("save.error.hint", this, null)
                .setParameters(getModelObject().getId(), ex.getMessage())
                .getString();
            log.error(msg);
            error(msg);
        }
    }

    private void doOnDelete() {
        final User user = getModelObject().toUser();
        final String userName = user.getUserName();
        try {
            userService.remove(user);
            info(new StringResourceModel("delete.successful.hint", this, null).setParameters(userName));
            setResponsePage(UserListPage.class);
        } catch (Exception ex) {
            String msg = new StringResourceModel("delete.error.hint", this, null)
                .setParameters(userName, ex.getMessage())
                .getString();
            log.error(msg);
            error(msg);
        }
    }

    public enum Mode {
        CREATE,
        MANAGE,
        EDIT,
        CHANGE_PASSWORD
    }
}

File: core/core-web/src/main/java/ch/difty/scipamato/core/web/user/UserEditPage.kt
package ch.difty.scipamato.core.web.user

import ch.difty.scipamato.common.logger
import ch.difty.scipamato.common.web.LABEL_RESOURCE_TAG
import ch.difty.scipamato.common.web.LABEL_TAG
import ch.difty.scipamato.core.auth.Role
import ch.difty.scipamato.core.auth.Roles
import ch.difty.scipamato.core.entity.User
import ch.difty.scipamato.core.entity.User.UserFields
import ch.difty.scipamato.core.persistence.OptimisticLockingException
import ch.difty.scipamato.core.persistence.UserService
import ch.difty.scipamato.core.web.CorePageParameters
import ch.difty.scipamato.core.web.common.BasePage
import de.agilecoders.wicket.core.markup.html.bootstrap.button.BootstrapButton
import de.agilecoders.wicket.core.markup.html.bootstrap.button.Buttons
import de.agilecoders.wicket.extensions.markup.html.bootstrap.confirmation.ConfirmationBehavior
import de.agilecoders.wicket.extensions.markup.html.bootstrap.form.select.BootstrapMultiSelect
import de.agilecoders.wicket.extensions.markup.html.bootstrap.form.select.BootstrapSelectConfig
import org.apache.wicket.AttributeModifier
import org.apache.wicket.authorization.UnauthorizedInstantiationException
import org.apache.wicket.authroles.authorization.strategies.role.annotations.AuthorizeInstantiation
import org.apache.wicket.bean.validation.PropertyValidator
import org.apache.wicket.markup.html.basic.Label
import org.apache.wicket.markup.html.form.CheckBox
import org.apache.wicket.markup.html.form.EmailTextField
import org.apache.wicket.markup.html.form.EnumChoiceRenderer
import org.apache.wicket.markup.html.form.Form
import org.apache.wicket.markup.html.form.PasswordTextField
import org.apache.wicket.markup.html.form.TextField
import org.apache.wicket.markup.html.form.validation.EqualPasswordInputValidator
import org.apache.wicket.markup.html.form.validation.IFormValidator
import org.apache.wicket.model.CompoundPropertyModel
import org.apache.wicket.model.Model
import org.apache.wicket.model.PropertyModel
import org.apache.wicket.model.StringResourceModel
import org.apache.wicket.request.mapper.parameter.PageParameters
import org.apache.wicket.spring.injection.annot.SpringBean
import org.apache.wicket.util.string.StringValue
import org.springframework.security.crypto.password.PasswordEncoder
import org.wicketstuff.annotation.mount.MountPath

private val log = logger()

/**
 * Loads the setting of the user with the id supplied in the page parameters.
 * If no valid integer is supplied in the page parameters, the settings of the
 * user in the current session is presented.
 *
 * The page can be opened in different modes:
 *
 *  * **CREATE:** The admin can create a new user
 *  * **MANAGE:** The admin edits other users accounts (including passwords)
 *  * **EDIT:** The active user can edit parts of his/her profile but cannot change the password.
 *  * **CHANGE_PASSWORD:** The (active) user can change the password but no other aspects of his/her profile
 */
@MountPath("user")
@AuthorizeInstantiation(Roles.USER, Roles.ADMIN)
class UserEditPage(pp: PageParameters) : BasePage<ChangePasswordUser>(pp) {

    private val mode: Mode

    @SpringBean
    private lateinit var passwordEncoder: PasswordEncoder

    @SpringBean
    private lateinit var userService: UserService

    init {
        mode = getModeFrom(pp)
        assertManageOrCreateModeForAdminOnly()
        initDefaultModel(pp)
    }

    private fun getModeFrom(pp: PageParameters): Mode {
        val modeValue = pp[CorePageParameters.MODE.getName()]
        return if (modeValue.isEmpty) Mode.EDIT else Mode.valueOf(modeValue.toString())
    }

    private fun assertManageOrCreateModeForAdminOnly() {
        if (isInAdminMode && !activeUser.roles.contains(Role.ADMIN)) {
            val msg = StringResourceModel("page.access-forbidden", this, null)
                .setParameters(activeUser.userName, this.javaClass).string
            log.warn(msg)
            throw UnauthorizedInstantiationException(pageClass)
        }
    }

    private val isInAdminMode: Boolean
        get() = mode == Mode.MANAGE || mode == Mode.CREATE
    private val isInUserOwnedMode: Boolean
        get() = !isInAdminMode
    private val isInPasswordResetMode: Boolean
        get() = mode == Mode.CHANGE_PASSWORD

    private fun canSetPasswords(): Boolean = isInAdminMode || isInPasswordResetMode

    private fun initDefaultModel(pp: PageParameters) {
        val userId = pp[CorePageParameters.USER_ID.getName()]
        determineUserId(userId)
            ?.let { loadUserWithIdFromDb(it) }
            ?: run { defaultModel = Model.of(ChangePasswordUser()) }
    }

    /**
     * If the user id is supplied as page parameter, use its id (or that of the active user,
     * if the parameter is not a numeric value). Otherwise use the id of the current user,
     * unless if in CREATE Mode, then prepare for new user by returning null.
     */
    private fun determineUserId(userIdStringValue: StringValue): Int? {
        val activeUserId = activeUser.id!!
        return when (mode) {
            Mode.CREATE -> null
            Mode.MANAGE -> userIdStringValue.toInt(activeUserId)
            Mode.EDIT, Mode.CHANGE_PASSWORD -> activeUserId
        }
    }

    private fun loadUserWithIdFromDb(userId: Int) {
        userService.findById(userId)
            .ifPresent { u: User -> defaultModel = Model.of(ChangePasswordUser(u, !isInPasswordResetMode)) }
    }

    override fun onInitialize() {
        super.onInitialize()
        val form = Form("form", CompoundPropertyModel(model)).also { add(it) }
        TextField<String>(UserFields.USER_NAME.fieldName).apply { isEnabled = isInAdminMode }
            .also { queueFieldAndLabel(it, PropertyValidator<Any>()) }
        TextField<String>(UserFields.FIRST_NAME.fieldName).apply {
            isEnabled = !isInPasswordResetMode
        }.also { queueFieldAndLabel(it, PropertyValidator<Any>()) }
        TextField<String>(UserFields.LAST_NAME.fieldName).apply {
            isEnabled = !isInPasswordResetMode
        }.also { queueFieldAndLabel(it, PropertyValidator<Any>()) }
        EmailTextField(UserFields.EMAIL.fieldName).apply {
            isEnabled = !isInPasswordResetMode
        }.also { queueFieldAndLabel(it) }
        CheckBox(UserFields.ENABLED.fieldName).apply {
            isEnabled = isInAdminMode
        }.also { queueFieldAndLabel(it) }
        Label(ROLES + LABEL_TAG, StringResourceModel("$ROLES$LABEL_RESOURCE_TAG", this, null)).apply {
            isVisible = isInAdminMode
        }.also { queue(it) }
        BootstrapMultiSelect(ROLES, PropertyModel(model, ROLES), listOf(*Role.values()), EnumChoiceRenderer(this)).apply {
            with(BootstrapSelectConfig()
                .withMultiple(true)
                .withLiveSearch(true)
                .withLiveSearchStyle("startsWith"))
            add(AttributeModifier(AM_DATA_WIDTH, "auto"))
            isVisible = isInAdminMode
        }.also { queue(it) }
        Label("$ROLES_STRING$LABEL_TAG", StringResourceModel("$ROLES_STRING$LABEL_RESOURCE_TAG", this, null)).apply {
            isVisible = isInUserOwnedMode
        }.also { queue(it) }
        Label(ROLES_STRING, modelObject!!.rolesString).apply {
            isVisible = isInUserOwnedMode
        }.also { queue(it) }
        PasswordTextField("currentPassword").apply {
            isRequired = isInPasswordResetMode
            isVisible = isInPasswordResetMode
            val currentPassword = this@UserEditPage.modelObject.password
            if (isInPasswordResetMode && currentPassword != null)
                add(CurrentPasswordMatchesValidator(passwordEncoder, currentPassword))
        }.also { queueFieldAndLabel(it) }
        val pw1Field = PasswordTextField(UserFields.PASSWORD.fieldName).apply {
            isRequired = isInPasswordResetMode
            isVisible = canSetPasswords()
        }.also { queueFieldAndLabel(it) }
        val pw2Field = PasswordTextField("password2").apply {
            isRequired = isInPasswordResetMode
            isVisible = canSetPasswords()
        }.also { queueFieldAndLabel(it) }

        form.add(EqualPasswordInputValidator(pw1Field, pw2Field) as IFormValidator)

        queueSubmitButton("submit")

        object : BootstrapButton("delete", StringResourceModel("delete.label"),
            Buttons.Type.Default) {
            override fun onSubmit() {
                super.onSubmit()
                doOnDelete()
            }
        }.apply {
            add(ConfirmationBehavior())
        }.also {
            queue(it)
        }
    }

    @Suppress("SameParameterValue")
    private fun queueSubmitButton(id: String) {
        queue(object : BootstrapButton(id, StringResourceModel("submit.label"), Buttons.Type.Default) {
            override fun onSubmit() {
                super.onSubmit()
                doOnSubmit()
            }
        })
    }

    private fun doOnSubmit() {
        if (modelObject != null) {
            try {
                val unsaved = modelObject.toUser()
                if (!canSetPasswords()) unsaved.password = null
                val user = userService.saveOrUpdate(unsaved)
                if (user != null) {
                    this@UserEditPage.modelObject = ChangePasswordUser(user)
                    if (mode == Mode.CHANGE_PASSWORD) {
                        info(StringResourceModel("password.changed.hint", this, this@UserEditPage.model).string)
                    } else {
                        info(StringResourceModel("save.successful.hint", this, null)
                            .setParameters(modelObject.id, user.userName).string)
                    }
                } else {
                    error(StringResourceModel("save.error.hint", this, null)
                        .setParameters(modelObject.id, "").string)
                }
            } catch (ole: OptimisticLockingException) {
                val msg = StringResourceModel("save.optimisticlockexception.hint", this, null)
                    .setParameters(ole.tableName, modelObject.id).string
                log.error(msg)
                error(msg)
            } catch (ex: Exception) {
                val msg = StringResourceModel("save.error.hint", this, null)
                    .setParameters(modelObject.id, ex.message).string
                log.error(msg)
                error(msg)
            }
        }
    }

    private fun doOnDelete() {
        val user = modelObject!!.toUser()
        val userName = user.userName
        try {
            userService.remove(user)
            info(StringResourceModel("delete.successful.hint", this, null).setParameters(userName))
            setResponsePage(UserListPage::class.java)
        } catch (ex: Exception) {
            val msg = StringResourceModel("delete.error.hint", this, null)
                .setParameters(userName, ex.message).string
            log.error(msg)
            error(msg)
        }
    }

    enum class Mode {
        CREATE, MANAGE, EDIT, CHANGE_PASSWORD
    }

    companion object {
        private const val serialVersionUID = 1L
        private const val AM_DATA_WIDTH = "data-width"
        private const val ROLES = "roles"
        private const val ROLES_STRING = "rolesString"
    }
}
